lib/ansible/modules/source_control/gitlab_service.py

```diff
@@ -0,0 +1,377 @@
+'''
+
+
+def init_definitions():
+    sd = json.loads(RAW_SERVICES_DEFINITIONS)
+    for service_slug, params in sd.items():
+        for name, definition in params.items():
+            if 'type' in definition:
+                definition['type'] = bool if definition['type'] == 'bool' else int if definition['type'] == 'int' else definition['type']
+            if name in GitLabServices.credentialParams:
+                definition['no_log'] = True
+    return sd
+
+
+class GitLabServices(object):
+    HOOK_EVENTS = ['push', 'issues', 'confidential_issues', 'merge_requests', 'tag_push', 'note', 'confidential_note', 'job', 'pipeline', 'wiki_page']
+    credentialParams = ['password', 'token', 'api_key']
+
+    def __init__(self, module, git):
+        self._module = module
+
+    def update(self, project, remote, active, params, events):
+        diff = not self.equals(remote.attributes, active, params, events)
+        if diff:
+            # remote.active = active
+            for k, v in self.expandEvents(events).items():
+                setattr(remote, k, v)
+            # when saving, params does directly in the request
+            # (remember they get retrieved under the "property" key of the remote dict())
+            for k, v in params.items():
+                setattr(remote, k, v)
+            if self._module.check_mode:
+                self._module.exit_json(changed=True)
+            remote.save()
+            return remote
+        else:
+            return False
+
+    def asApiObject(self, active, params, events):
+        obj = {'active': active}
+        obj.update(self.expandEvents(events))
+        obj.update(params)
+        return obj
+
+    def expandEvents(self, events):
+        # python >= 2.7: return {value + '_events': bool(value in events) for value in self.HOOK_EVENTS}
+        v = {}
+        for value in self.HOOK_EVENTS:
+            setattr(v, value + '_events', bool(value in events))
+        return v
+
+    def eventsEqual(self, obj, events):
+        return all((k in obj and obj[k] == v) for k, v in events.items())
+
+    def hasCredentials(self, attr):
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def equals(self, prev_attr, active, params, events):
+        # python >= 2.7: filtered_params = {k: v for k, v in params.items() if k not in self.credentialParams}
+        filtered_params = {}
+        for k, v in params.items():
+            if k not in self.credentialParams:
+                setattr(filtered_params, k, v)
+        # ToDo: this is not (yet?) supported by GitLab
+        # if prev_attr['active'] != active:
```

Comment:
Is it now ?

Reply:
still not, see https://gitlab.com/gitlab-org/gitlab-ce/issues/41113